lib/public_key/src/pubkey_cert.erl

```diff
@@ -558,82 +635,579 @@ root_cert(Name, Opts) ->
+                                             user_state = UserState}
+                     ) ->
+    case not (?anyPolicy == IssuerPolicy) andalso
+        not (?anyPolicy == SubjectPolicy) of
+        true ->
+            Tree = handle_policy_mapping_ext(Ext, Tree0,
+                                             PolicyMappingConstraint, AnyQualifiers),
+            {tree, Tree};
+        false ->
+            UserState = verify_fun(OtpCert, {bad_cert, {invalid_policy_mapping, Ext}},
+                                   UserState, VerifyFun),
+            {user_state, UserState}
+    end.
+
+%% 6.1.4. b continue:
+handle_policy_mapping_ext(#'PolicyMappings_SEQOF'{
+                         issuerDomainPolicy =
+                             IssuerPolicy},
+                         Tree0, 0, _) -> %% 6.1.4. b 2:
+    %% (2) If the policy_mapping variable is equal to 0:
+
+    %% (i) delete each node of depth i in the valid_policy_tree where
+    %% ID-P is the valid_policy.
+    %%
+    %% (ii) If there is a node in the valid_policy_tree of depth i-1
+    %% or less without any child nodes, delete that node.  Repeat this
+    %% step until there are no nodes of depth i-1 or less without
+    %% children.
+
+    Tree = pubkey_policy_tree:prune_leaves(Tree0, IssuerPolicy),
+    pubkey_policy_tree:prune_tree(Tree);
+handle_policy_mapping_ext(#'PolicyMappings_SEQOF'{
+                             issuerDomainPolicy = IssuerPolicy,
+                             subjectDomainPolicy = SubjectPolicy},
+                          Tree, N, AnyQualifiers) when N > 0 -> %% 6.1.4. b 1:
+    %% (1) If the policy_mapping variable is greater than 0, for each
+    %% node in the valid_policy_tree of depth i where ID-P is the
+    %% valid_policy, set expected_policy_set to the set of
+    %% subjectDomainPolicy values that are specified as equivalent to
+    %% ID-P by the policy mappings extension.
+
+    %% If no node of depth i in the valid_policy_tree has a
+    %% valid_policy of ID-P but there is a node of depth i with a
+    %% valid_policy of anyPolicy, then generate a child node of the
+    %% node of depth i-1 that has a valid_policy of anyPolicy as
+    %% follows:
+
+    %% (i) set the valid_policy to ID-P;
+
+    %% (ii) set the qualifier_set to the qualifier set of the policy
+    %% anyPolicy in the certificate policies extension of certificate
+    %% i; and
+
+    %% (iii) set the expected_policy_set to the set of
+    %% subjectDomainPolicy values that are specified as equivalent to
+    %% ID-P by the policy mappings extension.
+
+    MapPolicy =
+        fun(#{valid_policy := ValidPolicy, expected_policy_set := Set} = Node)
+              when ValidPolicy == IssuerPolicy ->
+                case Set == [ValidPolicy] of
+                    true ->
+                        Node#{expected_policy_set => [SubjectPolicy]};
+                    false ->
+                        Node#{expected_policy_set => Set ++ [SubjectPolicy]}
```

Comment:
We will iterate over a list of policy mappings that may produce  a  mapping value each. If we do not accumulate the mapped values we will always end up with only the last mapped value.

Reply:
> We will iterate over a list of policy mappings that may produce a mapping value each.
If we do not accumulate the mapped values we will always end up with only the last mapped value.

I agree accumulation is wanted, but I don't think `ValidPolicy` is in general expected in expected_policy_set after mapping is done. (Unless some mapping defines explicitly `ValidPolicy` as a target in SubjectPolicy).

I'm looking at RFC 5280 4.2.1.5:
_In general, certificate policies that appear in the issuerDomainPolicy field of the policy mappings extension are not considered acceptable policies for inclusion in subsequent certificates in the certification path. ..._

Current code handles that, but only if Set is `[ValidPolicy]` - what if it is 2 element list containing ValidPolicy? Can this situation be ruled out? Don't we need to delete ValidPolicy from expected policy set before processing mappings?

IMHO, this code will also spoil p1->p1 mapping - it will remove mapping ordered by CA.

Reply:
@IngelaAndin - have a 2nd look pls

Reply:
New solution!

Reply:
After discussion, I'm convinced that we could assume single element list is received before mapping starts.

we can get redundant data when p1->p1 mapping is 1st on list ...?

Reply:
@u3s new solution please check

Reply:
looks ok for me. can be resolved.